### PR TITLE
fix: updates RFC 0011 reflecting the current state of the audit scanner.

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Updates the RFC 0011 describing how audit scanner performs the scans to reflect the latest changes done in this component.

Related to https://github.com/kubewarden/audit-scanner/issues/209

### CODE REVIEW

File: rfc/0011-audit-checks.md [hunk 1]
```diff
@@ -113,6 +113,370 @@ resources of a given Namespace.
...
+target. The `GroupVersionResource` identifier helps in recognizing the specific
+Kubernetes resource types involved.
+
+**Note**: Only policies that review CREATE events are taken into account.
+
+#### Step 2: fetch Kubernetes resources to be audited
+
+This step mirrors the process described in Step 2 for the ClusterPolicyReport.
+It involves fetching the instances of Kubernetes resources present within the
+namespace from the API server, likely in a paginated manner to manage large
```
Comment (fabriziosestito) [inline rfc/0011-audit-checks.md]: remove likely, we are paginating now

File: rfc/0011-audit-checks.md [hunk 2]
```diff
@@ -113,6 +113,370 @@ resources of a given Namespace.
...
+load testing data, one single Policy Server running with 2 workers can evaluate
+about 1600 requests per second. Based on this data, we think the requests
+generated by the audit checks should not clog the Policy Server. If that turns
+out to be true, we can embrace different strategies to mitigate that.
+
+Until we have real world data, it's not useful to eagerly optimize this aspect.
+
+### Step 4: write `ClusterPolicyReport`
+
+The code will overwrite the old `ClusterPolicyReport` instance with a new generated
```
Comment (fabriziosestito) [inline rfc/0011-audit-checks.md]: We don't overwrite but we update the old object field with new data, also I would remove "data kept in memory" as we are keeping the data just for enough time to build the report patch.

File: rfc/0011-audit-checks.md [hunk 3]
```diff
@@ -113,6 +113,370 @@ resources of a given Namespace.
...
+the criteria set out in the policies, specifically focusing on the labels
+required by the policy. If a resource fits the criteria, the scanner simulates
+a `CREATE` event, incorporating the audited Kubernetes object as `request.object`
+within this event.
+
+### Step 3: performing evaluations
+
+Initially, the evaluation is executed through an HTTP request directed at the
+Policy Server that enforces the policy. This step is crucial for determining
+the compliance of each resource with the established cluster policies.
```
Comment (fabriziosestito) [inline rfc/0011-audit-checks.md]: I would add that we are cylcing all the resources and auditing the resources one by one against the policies they are targeted by.

File: rfc/0011-audit-checks.md [hunk 4]
```diff
@@ -72,11 +72,11 @@ equal to some hard coded values.
 If a policy makes use of username/groups at evaluation time, this policy will not
 produce meaningful evaluation results during the audit checks.
 
-The policy author should make this clear by adding the following annotation
+The policy author should make this clear by adding the following field
 inside of the `metadata.yml` file of the policy:
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
in the `metadata.yml` file of the policy:
```

File: rfc/0011-audit-checks.md [hunk 5]
```diff
@@ -113,6 +113,371 @@ resources of a given Namespace.
 
 That's because:
 
+* Once all the policies and involved resources are audited, the
+  `PolicyReport` audit is going to be overwritten with the results of the audit
+
+Users can still get live data about the rejections/approvals by looking both
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
Users can still get live data about rejections and approvals by looking both
```

File: rfc/0011-audit-checks.md [hunk 6]
```diff
@@ -113,6 +113,371 @@ resources of a given Namespace.
...
+* Policies in `monitor` mode should still behave like the ones in `protect`
+  mode. They have to produce rejections if needed
+
+By issuing the audit request against a real Policy Server, we solve these
+problems:
+
+* Policies can be hosted on registries secured with credentials. We don't have
+  to worry about being able to fetch the WebAssembly modules from these
+  authenticated registries
+* Policies might be hosted on insecure registires/registires with custom TLS
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
* Policies might be hosted on insecure registries or registries with custom TLS
```

File: rfc/0011-audit-checks.md [hunk 7]
```diff
@@ -113,6 +113,371 @@ resources of a given Namespace.
...
+* Some data used by the policy (like external OCI manifests, sigstore
+  verification results) can be cached by the Policy Server and reused between
+  the audit evaluations and the regular validations requests
+
+Sending all our auditing requests to the actual Policy Server could lead to
+poor performance of the regular admission requests. Currently, based on our
+load testing data, one single Policy Server running with 2 workers can evaluate
+about 1600 requests per second. Based on this data, we think the requests
+generated by the audit checks should not clog the Policy Server. If that turns
+out to be true, we can embrace different strategies to mitigate that.
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
out to be false, we can embrace different strategies to mitigate that.
```

File: rfc/0011-audit-checks.md [hunk 8]
```diff
@@ -113,6 +113,371 @@ resources of a given Namespace.
...
+The `/audit` API will behave in a slightly different way compared to the current `/validate`
+endpoint. The differences have been illustrated by the previous section of things
+RFC.
+
+## Changes to ClusterAdmissionPolicy and AdmissionPolicy CRDs
+
+It's necessary a change in the policy CRDs to mark a policy to be run in the background audit checks.
+The field would be called `backgroundAudit` and its default value will be `true`.
+Therefore, all the policies will run in the audit checks by default, unless the user
+configure the other way. Example of policies CRDs with the `backgroundAudit` field:
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
configures it the other way. Example of policies CRDs with the `backgroundAudit` field:
```

File: rfc/0011-audit-checks.md [hunk 9]
```diff
@@ -113,6 +113,371 @@ resources of a given Namespace.
...
+The audit scanner is a CLI program running in a non interactive way. It will be
+ran via a Kubernetes CronJob every 30 minutes. In the future we can provide a
+way to allow the user to configure this interval.
+
+In the initial version, the scanner will iterate sequentially over each
+namespace to generate its `PolicyReport` object for each audited resource. In
+the future we might want to change this implementation to scan the namespaces
+in parallel.
+
+By default, the scanner will look at all the namespaces defined inside of the
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
By default, the scanner will look at all the namespaces defined in the
```

File: rfc/0011-audit-checks.md [hunk 10]
```diff
@@ -113,6 +113,371 @@ resources of a given Namespace.
...
+the user to explicitly pick the namespaces to inspect. This could be used to
+provide a way to our user to generate the audit report of a specified namespace
+on demand, without having to wait for the next run on the scanner.
+
+
+# Drawbacks
+
+No real time audit info are provided by this solution. Users have to rely on
+Prometheus metrics and Open Telemetry trace events to have live data coming
+from the cluster.
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
No real time audit information is provided by this solution. Users have to rely on
Prometheus metrics and Open Telemetry trace events to have live data coming
from the cluster.
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
A few suggestions are marked up. One typo as well.

File: rfc/0011-audit-checks.md [hunk 11]
```diff
@@ -36,10 +36,10 @@ As a Kubernetes developer, I want access to the audit checks results to allow me
 # Detailed design
 
 The audit check will run at specified intervals and will produce audit reports using the `PolicyReport` format.
-There is going to be one `PolicyReport` object per namespace, plus one `ClusterPolicyReport` that will hold information
-about the cluster-wide resources.
+There is going to be one `PolicyReport` object per namespaced resource. As well as a `ClusterPolicyReport` 
+for each cluster wide resource 
 
-Details about the `PolicyReport` format can be found inside of this [RFC](https://github.com/kubewarden/rfc/pull/13/files).
+Details about the `PolicyReport` format can be found inside of this [RFC](https://github.com/kubewarden/rfc/blob/main/rfc/0012-policy-report.md).
```
Comment (jhkrug) [inline rfc/0011-audit-checks.md]:
```suggestion
Details about the `PolicyReport` format can be found in this [RFC](https://github.com/kubewarden/rfc/blob/main/rfc/0012-policy-report.md).
```

**Review by jhkrug — CHANGES_REQUESTED** [review]
One more 'inside of' change to 'in'. Sorry, missed it the first time.

**Review by flavio — APPROVED** [review]

**Review by jhkrug — APPROVED** [review]
lgtm

**Review by viccuad — APPROVED** [review]
LGTM.

**Pull request merged**